hi-ml-multimodal/src/health_multimodal/image/model/transformer.py

```diff
@@ -0,0 +1,262 @@
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  -------------------------------------------------------------------------------------------
+
+import math
+from dataclasses import dataclass
+from functools import partial
+from typing import Any, Callable, Optional, Set, Tuple
+
+import torch
+import torch.nn as nn
+from timm.models.layers import DropPath, Mlp, trunc_normal_
+from transformers.pytorch_utils import torch_int_div
+
+
+@dataclass
+class MultiHeadAttentionOutput:
+    mha_output: torch.Tensor
+    attention: Optional[torch.Tensor] = None
+
+
+class VisionTransformerPooler(nn.Module):
+    """
+    :param input_dim: Input feature dimension (i.e., channels in old CNN terminology)
+    :param grid_shape: Shape of the grid of patches per image
+    :param num_heads: Number of self-attention heads within the MHA block
+    :param num_blocks: Number of blocks per attention layer
+    :param norm_layer: Normalisation layer
+
+    `self.type_embed`: Is used to characterise prior and current scans, and
+                       create permutation variance across modalities/series.
+    `self.cls_token`: Cane be used to aggregate information across modalities and image series.
+    """
+
+    def __init__(self,
+                 input_dim: int,
+                 grid_shape: Tuple[int, int],
+                 num_heads: int = 8,
+                 num_blocks: int = 3,
+                 norm_layer: Any = partial(nn.LayerNorm, eps=1e-6)):
+        super().__init__()
+
+        block_kwargs = dict(dim=input_dim, num_heads=num_heads, mlp_ratio=1., drop=0.10, attn_drop=0.10,
+                            drop_path=0.25, act_layer=nn.GELU, norm_layer=norm_layer)
+        self.blocks = nn.ModuleList([Block(**block_kwargs) for _ in range(num_blocks)])
+        self.norm_post = norm_layer(input_dim)
+        self.use_cls_token = False
+        self.grid_shape = grid_shape
+        self.num_patches = grid_shape[0] * grid_shape[1]
+        self.num_blocks = num_blocks
+
+        # CLS token and type embeddings
+        num_series: int = 2
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, input_dim))
+        self.type_embed = nn.Parameter(torch.zeros(num_series, 1, input_dim))
+        trunc_normal_(self.cls_token, std=.02)
+        trunc_normal_(self.type_embed, std=.02)
+
+        # Positional embeddings 1 x L x C (L: Sequence length, C: Feature dimension)
+        self.pos_drop = nn.Dropout(p=0.10)
+        pos_embed_class = SinePositionEmbedding(embedding_dim=input_dim // 2, normalize=True)
+        pos_embed = pos_embed_class(mask=torch.ones([1, grid_shape[0], grid_shape[1]]))  # 1 x L x C
+        self.register_buffer("pos_embed", pos_embed, persistent=False)
+
+        # Initialisation
+        self.apply(self._init_weights)
+
+    def no_weight_decay(self) -> Set[str]:
+        return {'type_embed', 'cls_token'}
```

Comment:
Do we need this? It's currently not covered.

Reply:
Updated - We will need to update the checkpoint as well, but I think it was a good idea since the corresponding parameters were not trained.